### PR TITLE
[Sparse] Support SpMM and SDDMM

### LINKED ISSUE #4855: [Sparse] Support sparse-dense matmul and SDDMM
## 🔨Work Item

Project tracker: https://github.com/orgs/dmlc/projects/2

### PR DESCRIPTION
**Pull request by czkkkkkk**
## Description
Resolve #4855.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by czkkkkkk** [comment]
@jermainewang I only added the transpose flag in `SpMMImpl`, which is enough to avoid transposition in the backward. I am not sure whether we need the flag in the `SpMM`.

File: dgl_sparse/src/spmm.cc [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+  static tensor_list backward(AutogradContext* ctx, tensor_list grad_outputs);
+};
+
+void _SpMMSanityCheck(
+    c10::intrusive_ptr<SparseMatrix> sparse_mat, torch::Tensor sparse_val,
+    torch::Tensor dense_mat) {
+  const auto& sparse_mat_shape = sparse_mat->shape();
+  auto val_shape = sparse_val.sizes();
+  auto dense_shape = dense_mat.sizes();
+  CHECK(sparse_mat_shape[1] == dense_shape[0]);
```
Comment (jermainewang) [inline dgl_sparse/src/spmm.cc:37]: Add proper error messages for these checks?

File: dgl_sparse/src/spmm.cc [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+  return ret;
+}
+
+torch::Tensor SpMMAutoGrad::forward(
+    AutogradContext* ctx, c10::intrusive_ptr<SparseMatrix> sparse_mat,
+    torch::Tensor sparse_val, torch::Tensor dense_mat) {
+  _SpMMSanityCheck(sparse_mat, sparse_val, dense_mat);
+  auto ret = SpMMImpl(sparse_mat, sparse_val, dense_mat, false);
+
+  bool sparse_require_grad = sparse_val.requires_grad();
```
Comment (jermainewang) [inline dgl_sparse/src/spmm.cc:107]: const

File: dgl_sparse/src/spmm.cc [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+  bool dense_require_grad = dense_mat.requires_grad();
```
Comment (jermainewang) [inline dgl_sparse/src/spmm.cc:108]: const

File: dgl_sparse/src/sddmm.cc [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  if (sparse_mat->HasCSR() || !sparse_mat->HasCOO()) {
+    fmt = SparseFormat::kCSR;
+  } else {
+    fmt = SparseFormat::kCOO;
+  }
+  if (mat2_tr.dim() == 1) {
+    mat1 = mat1.view({-1, 1});
+    mat2_tr = mat2_tr.view({-1, 1});
+  }
+  int64_t out_row = sparse_mat->nnz();
```
Comment (jermainewang) [inline dgl_sparse/src/sddmm.cc:33]: const

File: include/dgl/kernel.h [hunk 5]
```diff
@@ -42,10 +42,12 @@ void SpMM(
...
  * @param vfeat The destination node feature.
  * @param out The output feature on edge.
+ * @param lhs_target Type of `ufeat`. (0: source; 1: edge; 2: destination)
+ * @param rhs_target Type of `vfeat`. (0: source; 1: edge; 2: destination)
  */
 void SDDMM(
-    const std::string& op, HeteroGraphPtr graph, NDArray ufeat, NDArray efeat,
-    NDArray out);
+    const std::string& op, HeteroGraphPtr graph, NDArray ufeat, NDArray vfeat,
+    NDArray out, int lhs_target, int rhs_target);
```
Comment (jermainewang) [inline include/dgl/kernel.h:50]: Is this change still needed?

File: python/dgl/mock_sparse2/matmul.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    ----------
+    A : SparseMatrix or DiagMatrix
+        Sparse matrix of shape (N, M) with values of shape (nnz)
+    X : torch.Tensor
+        Dense tensor of shape (M, F) or (M)
+
+    Returns
+    -------
+    torch.Tensor
+        The result of multiplication
```
Comment (jermainewang) [inline python/dgl/mock_sparse2/matmul.py:27]: Missing result shape.

File: python/dgl/mock_sparse2/matmul.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+from .diag_matrix import DiagMatrix
+
+from .sparse_matrix import SparseMatrix
+
+__all__ = ["spmm"]
+
+
+def spmm(A: Union[SparseMatrix, DiagMatrix], X: torch.Tensor) -> torch.Tensor:
+    """Multiply a sparse matrix by a dense matrix
```
Comment (jermainewang) [inline python/dgl/mock_sparse2/matmul.py:15]:
```suggestion
    """Multiply a sparse matrix by a dense matrix.
```

File: python/dgl/mock_sparse2/matmul.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+    Examples
+    --------
+
+    >>> row = torch.tensor([0, 1, 1])
+    >>> col = torch.tensor([1, 0, 1])
+    >>> val = torch.randn(len(row))
+    >>> A = create_from_coo(row, col, val)
+    >>> X = torch.randn(2, 3)
+    >>> result = A @ X
```
Comment (jermainewang) [inline python/dgl/mock_sparse2/matmul.py:37]: `dgl.sparse.spmm` not the syntax sugar `@`.

File: python/dgl/mock_sparse2/matmul.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    assert isinstance(
+        X, torch.Tensor
+    ), f"Expect arg2 to be a torch.Tensor, got {type(X)}"
+    assert (
+        A.shape[1] == X.shape[0]
+    ), f"Expect arg1.shape[1] == arg2.shape[0], got {A.shape[1]} and {X.shape[0]}"
+    val_dim = len(A.val.shape)
+    assert val_dim == 1, f"Expect arg1.val to be a 1D tensor, got {val_dim}D"
+    val_dim = len(X.shape)
+    assert val_dim <= 2, f"Expect arg2 to be a 1D/2D tensor, got {val_dim}D"
```
Comment (jermainewang) [inline python/dgl/mock_sparse2/matmul.py:55]: I saw you've checked those shapes in `_SpMMSanityCheck` so you could omit them in python.

File: python/dgl/mock_sparse2/matmul.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+    if not isinstance(A, SparseMatrix):
```
Comment (jermainewang) [inline python/dgl/mock_sparse2/matmul.py:57]: Add a comment `The input is DiagMatrix. Cast it to SparseMatrix.`.

File: tests/pytorch/mock_sparse2/test_mm.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+    XX = X.clone().detach()
+    XX.requires_grad_()
+    dense_result = torch.sparse.mm(adj, XX)
+    dense_result.backward(grad)
+    assert torch.allclose(sparse_result, dense_result)
+    assert torch.allclose(X.grad, XX.grad)
+    assert torch.allclose(adj.grad.coalesce().values(), val.grad)
+
+
+def test_spmm_coo_one_dim_rhs():
```
Comment (jermainewang) [inline tests/pytorch/mock_sparse2/test_mm.py:48]: Why no `test_spmm_csr|csc_one_dim_rhs()`?

**Review by jermainewang — CHANGES_REQUESTED** [review]

File: dgl_sparse/src/spmm.cc [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+    cache_dense_mat = dense_mat;
+  }
+  ctx->saved_data["sparse_matrix"] = sparse_mat;
+  ctx->saved_data["sparse_require_grad"] = sparse_require_grad;
+  ctx->saved_data["dense_require_grad"] = dense_require_grad;
+  ctx->save_for_backward({cache_sparse_val, cache_dense_mat});
+  return ret;
+}
+
+tensor_list SpMMAutoGrad::backward(
```
[the remark this answers is not in the document]
Comment (czkkkkkk) [reply dgl_sparse/src/spmm.cc:122]: Currently no. Will file an issue.

File: dgl_sparse/src/sddmm.cc [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+    aten::COOSDDMM(
+        op.c_str(), coo, dgl_mat1, dgl_mat2_tr, dgl_ret, 0 /* Lhs target: u */,
+        2 /* rhs target: v */);
+  }
+  return ret;
+}
+
+torch::Tensor SDDMMAutoGrad::forward(
+    AutogradContext* ctx, const c10::intrusive_ptr<SparseMatrix>& sparse_mat,
+    torch::Tensor mat1, torch::Tensor mat2) {
```
Comment (jermainewang) [inline dgl_sparse/src/sddmm.cc:78]: If SDDMM is formulated as `(mat1 @ mat2) * sparse_mat`, then it should multiply the nonzero values of the sparse matrix too (and therefore need to support autograd of the nonzero values).

File: dgl_sparse/src/spmm.cc [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+    } else {  // COO
+      // Use the reverse order of aten::COOSpMM because it calculates A^T @ X.
+      auto coo = COOToOldDGLCOO(sparse_mat->COOPtr());
+      coo = aten::COOTranspose(coo);
+      aten::COOSpMM(
+          op.c_str(), reduce.c_str(), coo, dgl_dense_mat, dgl_sparse_val,
+          dgl_ret, {});
+    }
+  } else {  // transpose_sparse
+    if (sparse_mat->HasCSC() || !sparse_mat->HasCOO()) {
```
Comment (jermainewang) [inline dgl_sparse/src/spmm.cc:84]: Add a comment "Use CSC in DGL's CSRSpMM is equivalent as computing A^T @ X."

**Review by jermainewang — CHANGES_REQUESTED** [review]

File: dgl_sparse/include/sparse/sddmm.h [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+ *
+ * This function does not take care of autograd.
+ *
+ * @param sparse_mat The sparse matrix.
+ * @param mat1 The first dense matrix.
+ * @param mat2_tr Transposition of the second matrix.
+ *
+ * @return Dense tensor.
+ */
+torch::Tensor SDDMMImpl(
```
[the remark this answers is not in the document]
Comment (frozenbugs) [reply dgl_sparse/include/sparse/sddmm.h:36]: We should never expose implementation details (local method), keep implementation details in the cc file, only expose the public APIs.

File: dgl_sparse/src/sddmm.cc [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+    torch::Tensor mat2) {
+  CHECK_EQ(mat1.dim(), 2)
+      << "SDDMM: the first dense matrix should be two-dimensional.";
+  CHECK_EQ(mat2.dim(), 2)
+      << "SDDMM: the second dense matrix should be two-dimensional.";
+  CHECK_EQ(sparse_mat->shape()[0], mat1.size(0))
+      << "SDDMM: the first dense matrix should have the same first dimension "
+         "as the sparse matrix";
+  CHECK_EQ(sparse_mat->shape()[1], mat2.size(1))
+      << "SDDMM: the seond dense matrix should have the same second dimension "
```
Comment (frozenbugs) [inline dgl_sparse/src/sddmm.cc:41]: typo: the second dense matrix..

File: dgl_sparse/src/sddmm.cc [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+  const int64_t out_row = sparse_mat->nnz();
+  const std::vector<int64_t> shape({out_row});
+  auto ret = torch::zeros(shape, mat1.options());
+  const std::string op = "dot";
+  auto dgl_mat1 = TorchTensorToDGLArray(mat1);
+  auto dgl_mat2_tr = TorchTensorToDGLArray(mat2_tr);
+  auto dgl_ret = TorchTensorToDGLArray(ret);
+  // Use CSR if the sparse matrix has CSR or does not have COO. Otherwise use
+  // COO.
+  if (sparse_mat->HasCSR() || !sparse_mat->HasCOO()) {
```
Comment (frozenbugs) [inline dgl_sparse/src/sddmm.cc:62]: Will this crash if the !sparse_mat->HasCSR() && !sparse_mat->HasCOO()?

Comment (frozenbugs) [reply dgl_sparse/src/sddmm.cc:62]: If the sparse_mat should either HasCSR() or  HasCOO(). Then do:
If (sparse_mat->HasCSR()) {
} else if (sparse_mat->HasCOO()) {
} else {
    throw ...
}

File: dgl_sparse/src/sddmm.cc (hunk 13 above)
Comment (frozenbugs) [reply dgl_sparse/src/sddmm.cc:78]: From the doc https://docs.dgl.ai/en/latest/generated/dgl.mock_sparse.sddmm.html#dgl.mock_sparse.sddmm.
SDDMM is defined as (B @ C) * dglsp.ones_like(A) + A.

File: dgl_sparse/include/sparse/sddmm.h [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+    const c10::intrusive_ptr<SparseMatrix>& sparse_mat, torch::Tensor mat1,
+    torch::Tensor mat2_tr);
+
+/**
+ * @brief Perform a sampled matrix multiplication of a sparse matrix and two
+ * dense matrices. If the sparse matrix has shape (n, m), `mat1` and `mat2` must
+ * have shapes of `(n, k)` and `(k, m)` or `(n,)` and `(m,)` respectively. And
+ * the returned tensor has shape `(sparse_matrix->nnz(),)`.
```
Comment (frozenbugs) [inline dgl_sparse/include/sparse/sddmm.h:44]: Probably worth to add mathematically formula here to help users to understand definition of sddmm.

File: tests/pytorch/mock_sparse2/test_mm.py [hunk 19]
```diff
@@ -0,0 +1,51 @@
+import sys
```
Comment (frozenbugs) [inline tests/pytorch/mock_sparse2/test_mm.py:1]: test_mm.py -> test_matmul.py

File: dgl_sparse/include/sparse/sddmm.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (frozenbugs) [reply dgl_sparse/include/sparse/sddmm.h:36]: It is actually surprising that we need SDDMMImpl in dgl_sparse/src/spmm.cc.

**Comment by frozenbugs** [comment]
nit: This PR is a little bit too long, worth to break into 2 PRs, one implement SDDMM and SpMM.

One good reason to do so is, while reading this PR, it is very easy to neglect the fact that SDDMMimpl is used in SpMM.cc. While breaking it to 2 PRs, it would be very obvious, since you need to expose the SDDMMimpl in the second PR in sddmm.h.

File: dgl_sparse/include/sparse/sddmm.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (czkkkkkk) [reply dgl_sparse/include/sparse/sddmm.h:36]: I see. But we still need headers `dgl_sparse/src/spmm_impl.h` and `dgl_sparse/src/sddmm_impl.h` because the backward functions of `SpMM` and `SDDMM` depend on `SDDMMImpl` and `SpMMImpl` respectively.

File: dgl_sparse/src/sddmm.cc (hunk 13 above)
Comment (czkkkkkk) [reply dgl_sparse/src/sddmm.cc:78]: That is not `SDDMM` but `sampled_addmm` following Pytorch (https://pytorch.org/docs/stable/generated/torch.sparse.sampled_addmm.html). I think the most common used `SDDMM` is `(B @ C) * dglsp.ones_like(A) * A` (http://tensor-compiler.org/docs/machine_learning.html.)

File: dgl_sparse/src/sddmm.cc (hunk 17 above)
Comment (czkkkkkk) [reply dgl_sparse/src/sddmm.cc:62]: > Will this crash if the !sparse_mat->HasCSR() && !sparse_mat->HasCOO()?

It would not crash because it will convert CSC to CSR for computation. The logic here is to use CSR or COO if they exist. Otherwise convert CSC to CSR for computation.

**Comment by czkkkkkk** [comment]
> nit: This PR is a little bit too long, worth to break into 2 PRs, one implement SDDMM and SpMM.
> 
> One good reason to do so is, while reading this PR, it is very easy to neglect the fact that SDDMMimpl is used in SpMM.cc. While breaking it to 2 PRs, it would be very obvious, since you need to expose the SDDMMimpl in the second PR in sddmm.h.

I didn't get you. Do you mean in the first PR we put `SDDMMImpl` in the `spmm.cc`? And in the second PR, move it to `sddmm.h`?

**Review by jermainewang — APPROVED** [review]

File: dgl_sparse/src/sddmm.cc (hunk 13 above)
Comment (czkkkkkk) [reply dgl_sparse/src/sddmm.cc:51]: > If SDDMM is formulated as (mat1 @ mat2) * sparse_mat, then it should multiply the nonzero values of the sparse matrix too (and therefore need to support autograd of the nonzero values).

The multiply is handled by libtorch.

File: dgl_sparse/src/matmul.cc [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+      transpose_sparse ? sparse_mat->shape()[1] : sparse_mat->shape()[0];
+  const std::vector<int64_t> shape = {out_row, dense_mat.size(1)};
+
+  auto ret = torch::zeros(shape, dense_mat.options());
+  auto dgl_sparse_val = TorchTensorToDGLArray(sparse_val);
+  auto dgl_dense_mat = TorchTensorToDGLArray(dense_mat);
+  auto dgl_ret = TorchTensorToDGLArray(ret);
+  if (!transpose_sparse) {
+    if (sparse_mat->HasCSR() || !sparse_mat->HasCOO()) {
+      auto csr = CSRToOldDGLCSR(sparse_mat->CSRPtr());
```
Comment (frozenbugs) [inline dgl_sparse/src/matmul.cc:35]: As discussed, please add comment here.
// sparse_mat->CSRPtr() will implicitly convert CSC to CSR format.

File: dgl_sparse/src/matmul.cc [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+      // Use the reverse order of aten::COOSpMM because it calculates A^T @ X.
+      auto coo = COOToOldDGLCOO(sparse_mat->COOPtr());
+      coo = aten::COOTranspose(coo);
+      aten::COOSpMM(
+          op.c_str(), reduce.c_str(), coo, dgl_dense_mat, dgl_sparse_val,
+          dgl_ret, {});
+    }
+  } else {  // transpose_sparse
+    // Use CSC in DGL's CSRSpMM is equivalent as computing A^T @ X.
+    if (sparse_mat->HasCSC() || !sparse_mat->HasCOO()) {
```
Comment (frozenbugs) [inline dgl_sparse/src/matmul.cc:49]: ditto.

File: dgl_sparse/src/matmul.cc [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+      auto csc = CSRToOldDGLCSR(sparse_mat->CSCPtr());
```
Comment (frozenbugs) [inline dgl_sparse/src/matmul.cc:50]: ditto.

File: dgl_sparse/src/matmul.cc [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+  const int64_t out_row = sparse_mat->nnz();
+  const std::vector<int64_t> shape({out_row});
+  auto ret = torch::zeros(shape, mat1.options());
+  const std::string op = "dot";
+  auto dgl_mat1 = TorchTensorToDGLArray(mat1);
+  auto dgl_mat2_tr = TorchTensorToDGLArray(mat2_tr);
+  auto dgl_ret = TorchTensorToDGLArray(ret);
+  // Use CSR if the sparse matrix has CSR or does not have COO. Otherwise use
+  // COO.
+  if (sparse_mat->HasCSR() || !sparse_mat->HasCOO()) {
```
Comment (frozenbugs) [inline dgl_sparse/src/matmul.cc:77]: ditto

File: dgl_sparse/src/matmul.cc [hunk 24]
```diff
@@ -0,0 +1,92 @@
...
+  const int64_t out_row =
+      transpose_sparse ? sparse_mat->shape()[1] : sparse_mat->shape()[0];
+  const std::vector<int64_t> shape = {out_row, dense_mat.size(1)};
+
+  auto ret = torch::zeros(shape, dense_mat.options());
+  auto dgl_sparse_val = TorchTensorToDGLArray(sparse_val);
+  auto dgl_dense_mat = TorchTensorToDGLArray(dense_mat);
+  auto dgl_ret = TorchTensorToDGLArray(ret);
+  if (!transpose_sparse) {
+    if (sparse_mat->HasCSR() || !sparse_mat->HasCOO()) {
```
[the remark this answers is not in the document]
Comment (frozenbugs) [reply dgl_sparse/src/matmul.cc:36]: I do think you should put CSC in the list.
To emphasize that if both COO and CSC existing, COO is preferred.

File: python/dgl/mock_sparse2/matmul.py [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+    # The input is a DiagMatrix. Cast it to SparseMatrix
+    if not isinstance(A, SparseMatrix):
+        A = A.as_sparse()
+    return torch.ops.dgl_sparse.spmm(A.c_sparse_matrix, X)
+
+
+def mm_sp(
+    A1: SparseMatrix, A2: Union[torch.Tensor, SparseMatrix, DiagMatrix]
+) -> Union[torch.Tensor, SparseMatrix]:
+    """Internal function for multiplying a sparse matrix by a dense/sparse/diagonal matrix.
```
Comment (frozenbugs) [inline python/dgl/mock_sparse2/matmul.py:59]: nit: 80 chars.

File: python/dgl/mock_sparse2/matmul.py [hunk 26]
```diff
@@ -0,0 +1,101 @@
...
+    >>> A2 = torch.randn(2, 3)
+    >>> result = A1 @ A2
+    >>> print(type(result))
+    <class 'torch.Tensor'>
+    >>> print(result.shape)
+    torch.Size([2, 3])
+    """
+    assert isinstance(
+        A2, (torch.Tensor, SparseMatrix, DiagMatrix)
+    ), f"Expect arg2 to be a torch Tensor, SparseMatrix, or DiagMatrix object, got {type(A2)}"
```
Comment (frozenbugs) [inline python/dgl/mock_sparse2/matmul.py:93]: 80 chars.

File: dgl_sparse/src/sddmm.cc [hunk 27]
```diff
@@ -0,0 +1,103 @@
...
+    // SDDMM(M, A, B) = C. dB = SpMM(dC^T, A)^T
+    auto mat2_tr_grad = SpMMNoAutoGrad(sparse_mat, grad, mat1, true);
+    mat2_grad = mat2_tr_grad.transpose(0, 1).contiguous();
+  }
+  return {torch::Tensor(), mat1_grad, mat2_grad};
+}
+
+c10::intrusive_ptr<SparseMatrix> SDDMM(
+    const c10::intrusive_ptr<SparseMatrix>& sparse_mat, torch::Tensor mat1,
+    torch::Tensor mat2) {
```
[the remark this answers is not in the document]
Comment (czkkkkkk) [reply dgl_sparse/src/sddmm.cc:91]: It should. I missed it.

File: python/dgl/mock_sparse2/sddmm.py [hunk 28]
```diff
@@ -0,0 +1,54 @@
...
+        out = (mat1 @ mat2) * A
+
+    Parameters
+    ----------
+    A : SparseMatrix
+        Sparse matrix of shape `(M, N)`.
+    mat1 : Tensor
+        Dense matrix of shape `(M, K)`
+    mat2 : Tensor
+        Dense matrix of shape `(K, N)`
```
Comment (jermainewang) [inline python/dgl/mock_sparse2/sddmm.py:30]: From the test cases, it seems that you also support `mat1` and `mat2` being vectors? If so, you may want to extend the shape description here and the example section too.

Comment (jermainewang) [reply python/dgl/mock_sparse2/sddmm.py:30]: Feel free to fix it in another PR if you want to unblock the development of spspmm.

Comment (czkkkkkk) [reply python/dgl/mock_sparse2/sddmm.py:30]: OK. Will fix it in the SpSpMM PR.
